src/main/java/org/tarantool/TarantoolClientImpl.java

```diff
@@ -640,35 +690,64 @@ protected int getState() {
             return state.get();
         }
 
+        /**
+         * Set CLOSED state, drop RECONNECT state.
+         */
         protected boolean close() {
-            for (;;) {
+            for (; ; ) {
                 int st = getState();
+
+                /* CLOSED is the terminal state. */
                 if ((st & CLOSED) == CLOSED)
                     return false;
+
+                /* Drop RECONNECT, set CLOSED.
+                 *
+                 * XXX: Should not we also drop other states? Or reimplement
```

Comment:
Maybe now is really the time to make this helper fair FSM with explicit state diagram and clear API. Hope it will allow us to finally forget about such weird races.

Reply:
yes, it requires some reworking. We need to figure out a better design for API and internal implementation.

Reply:
Okay. Please, file an issue for refactor these things.